src/Microsoft.Sbom.Api/Workflows/SbomGenerationWorkflow.cs

```diff
@@ -94,38 +95,54 @@ public virtual async Task<bool> RunAsync()
-                            manifestInfo.Version,
-                            fileArrayGenerator,
-                            packageArrayGenerator,
-                            relationshipsArrayGenerator,
-                            externalDocumentReferenceGenerator).
-                            ConfigureAwait(false);
-
-                        // Write headers
-                        serializationStrategy.AddHeadersToSbom(sbomConfigs, config);
-
-                        // Finalize JSON
-                        config.JsonSerializer?.FinalizeJsonObject();
-                    }
+                        var strategy = JsonSerializationStrategyFactory.GetStrategy(config.ManifestInfo.Version);
+                        strategy.StartGraphArray(manifestInfosFromConfig, sbomConfigs);
+                    });
+
+                    // Write all the JSON documents from the generationResults to the manifest based on the manifestInfo.
+                    var fileGenerationResult = await fileArrayGenerator.GenerateAsync(manifestInfosFromConfig);
+
+                    var packageGenerationResult = await packageArrayGenerator.GenerateAsync(manifestInfosFromConfig);
+
+                    var externalDocumentReferenceGenerationResult = await externalDocumentReferenceGenerator.GenerateAsync(manifestInfosFromConfig);
 
-                    // Generate SHA256 for manifest json
-                    GenerateHashForManifestJson(config.ManifestJsonFilePath);
+                    var relationshipGenerationResult = await relationshipsArrayGenerator.GenerateAsync(manifestInfosFromConfig);
+
+                    // Concatenate all the errors from the generationResults.
+                    validErrors = validErrors.Concat(fileGenerationResult.Errors);
+                    validErrors = validErrors.Concat(packageGenerationResult.Errors);
+                    validErrors = validErrors.Concat(externalDocumentReferenceGenerationResult.Errors);
+                    validErrors = validErrors.Concat(relationshipGenerationResult.Errors);
+
+                    // Write metadata dictionary to SBOM. This is a no-op for SPDX 3.0 and above.
+                    ForEachManifestFromConfig(manifestInfosFromConfig, config =>
+                    {
+                        var strategy = JsonSerializationStrategyFactory.GetStrategy(config.ManifestInfo.Version);
+                        strategy.AddMetadataToSbom(sbomConfigs, config);
+                    });
+
+                    ForEachManifestFromConfig(manifestInfosFromConfig, config =>
+                    {
+                        var strategy = JsonSerializationStrategyFactory.GetStrategy(config.ManifestInfo.Version);
+                        strategy.EndGraphArray(manifestInfosFromConfig, sbomConfigs);
+                    });
+
+                    // Finalize JSON
+                    ForEachManifestFromConfig(manifestInfosFromConfig, config => config.JsonSerializer.FinalizeJsonObject());
                 }
```

Comment:
We eventually need to `Dispose` the `JsonDocument` objects after we finalize

Reply:
Implemented in my recent commits, where I dispose in the implementations of IJsonArrayGenerator